common/libs/framing/frame_detector.h

```diff
@@ -0,0 +1,206 @@
+class FrameDetector : public RxListener {
+public:
+  enum class State {
+    LOST, // Frame detector does not know where it is in the stream of
+          // bytes coming into RxBuffer
+    WAIT_FOR_START_MARKER, // Frame detector is waiting for a frame start marker
+    RECEIVING_FRAME        // Frame detector is receiving a frame
+  };
+
+  FrameDetector(RxBuffer &t) : rx_buffer_(t){};
+
+  // Starts frame detector
+  [[nodiscard]] bool Begin() {
+    state_ = State::LOST;
+    frame_available_ = false;
+    return rx_buffer_.Begin(this);
+  }
+
+  // Callback method called when RxBuffer is full
+  void OnRxComplete() override {
+    // We should never reach the full read of rx buffer.
+    // If we get here, this means, there are no marker
+    // chars in the stream, so we are lost
+    state_ = State::LOST;
+    RestartRX();
+  }
+
+  // Callback method called when RxBuffer receives a marker character;
+  // Note, MARK character is already written to RxBuffer so all conditions
+  // evaluating rx_buffer_.ReceivedLength() take that into account
+  void OnCharacterMatch() override {
+    marker_count_++;
+
+    switch (state_) {
+    case State::LOST:
+      if (IsDataReceived()) {
+        // if we have received something more than just markers,
+        // we assume, this is the frame end marker, so wait
+        // for start
+        state_ = State::WAIT_FOR_START_MARKER;
+        RestartRX();
+      } else if (IsOnlyMarkersReceived()) {
+        // We have received just this marker
+        // In this case, we were lucky to get to LOST state in the interframe
+        // silence, assume this is the start of the frame
+        state_ = State::RECEIVING_FRAME;
+      } else {
+        // Should never end up here
+        // DMA is not working?
+        // TODO alert, safe reset
+        RestartRX();
+      }
+      break;
+    case State::WAIT_FOR_START_MARKER:
+      if (IsDataReceived()) {
+        // some junk received while waiting for the start marker,
+        // but should have been just silence
+        state_ = State::LOST;
+        RestartRX();
+      } else if (IsOnlyMarkersReceived()) {
+        // we have received the marker we are waiting for
+        state_ = State::RECEIVING_FRAME;
+      } else {
+        // Should never end up here
+        // DMA is not working?
+        // TODO alert, safe reset
+        state_ = State::LOST;
+        RestartRX();
+      }
+      break;
+    case State::RECEIVING_FRAME:
+      // end marker received, check if we got something
+      if (IsDataReceived()) {
+        // yes, we got data, thus we got the frame we can pass further
+        processReceivedData();
+        state_ = State::WAIT_FOR_START_MARKER;
+        RestartRX();
+      } else if (IsOnlyMarkersReceived()) {
+        // repeated marker char received
+        // this means we have received a 0 lenght frame
+        // ignore it and continue receiving frame bytes
+        // rx_buffer_.RestartRX(this);
```

Comment:
Should this code be uncommented? If the code should be here, that means there was a bug, please add a test; otherwise delete the code.

Reply:
Sorry for this mishap. It should be uncommended. Fixed now.